source/simulator.hpp

```diff
@@ -10,42 +15,174 @@
 #include "device/sender.hpp"
 #include "device/switch.hpp"
 #include "link.hpp"
+#include "logger/logger.hpp"
+#include "scheduler.hpp"
 
 namespace sim {
 
+using routing_table_t =
+    std::unordered_map<std::shared_ptr<IRoutingDevice>, std::shared_ptr<ILink>>;
+
+static routing_table_t bfs(std::shared_ptr<IRoutingDevice>& start_device) {
+    routing_table_t routing_table;
+    std::queue<std::shared_ptr<IRoutingDevice>> queue;
+    std::set<std::shared_ptr<IRoutingDevice>> used;
+    queue.push(start_device);
+
+    while (!queue.empty()) {
+        std::shared_ptr<IRoutingDevice> device = queue.front();
+        queue.pop();
+        if (used.contains(device)) {
+            continue;
+        }
+        used.insert(device);
+        std::set<std::shared_ptr<ILink>> outlinks = device->get_outlinks();
+        for (std::shared_ptr<ILink> link : outlinks) {
+            auto next_hop = link->get_to();
+            auto curr_device = link->get_from();
+            if (used.contains(next_hop)) {
+                continue;
+            }
+            if (curr_device == start_device) {
+                routing_table[next_hop] = link;
+            } else if (!routing_table.contains(next_hop)) {
+                routing_table[next_hop] = routing_table[curr_device];
+            }
+            queue.push(next_hop);
+        }
+    }
+    return routing_table;
+};
+
+template <typename TSender, typename TSwitch, typename TReceiver,
+          typename TFlow>
```

Comment:
I think we also need template Link:
```suggestion
template <typename TSender, typename TSwitch, typename TReceiver,
          typename TFlow, typename TLink>
```